lib/internal/test_runner/test.js

```diff
@@ -176,22 +176,118 @@ function testMatchesPattern(test, patterns) {
 }
 
 class TestPlan {
-  constructor(count) {
-    validateUint32(count, 'count');
+  #waitIndefinitely = false;
+
+  constructor(count, options = kEmptyObject) {
+    this.#validateConstructorArgs(count, options);
     this.expected = count;
     this.actual = 0;
+    this.#setWaitOption(options);
+  }
+
+  #validateConstructorArgs(count, options) {
+    validateUint32(count, 'count');
+    validateObject(options, 'options');
+  }
+
+  #setWaitOption(options) {
+    switch (typeof options.wait) {
+      case 'boolean':
+        this.wait = options.wait;
+        this.#waitIndefinitely = options.wait;
+        break;
+      case 'number':
+        validateNumber(options.wait, 'options.wait', 0, TIMEOUT_MAX);
+        this.wait = options.wait;
+        break;
+      default:
+        if (options.wait !== undefined) {
+          throw new ERR_INVALID_ARG_TYPE('options.wait', ['boolean', 'number'], options.wait);
+        }
+    }
+  }
+
+  #planMet() {
+    return this.actual === this.expected;
+  }
+
+  #startPoller() {
```

Comment:
I don't think we need a poller. `check()` should only need to create a timeout (if needed). The promise created here can be a property on the `TestPlan` object. Then, `count()` can check if that promise has been created. If it has been created and the plan is met, then `count()` resolves the promise. Otherwise, the timeout will elapse at some point and reject the promise.

Reply:
Yes, it makes more sense and it's definitely more efficient!
I'll update the implementation following your suggestion!

Reply:
done